include/marisa/exception.h

```diff
@@ -19,7 +19,7 @@ class Exception : public std::exception {
   Exception(const Exception &ex)
       : std::exception(), filename_(ex.filename_), line_(ex.line_),
         error_code_(ex.error_code_), error_message_(ex.error_message_) {}
-  virtual ~Exception() throw() {}
+  virtual ~Exception() noexcept;
```

Comment:
`noexcept` is C++11. I seem to recall that @s-yata likes to keep this compatible with C++03, so maybe add some `#if __cplusplus >= ...`?

Reply:
I believe that throw() is also C++11. So I don't think this is changing the C++ level. It's also worth mentioning that throw() was deprecated in C++17 and removed in C++20. So there isn't a good reason to state that a destructor should throw exceptions.

https://en.cppreference.com/w/cpp/language/noexcept_spec

Do you still want this #if'd?

Reply:
> I believe that throw() is also C++11. 

It looks like it's C++98.  See `exception-specification`: https://www.lirmm.fr/~ducour/Doc-objets/ISO+IEC+14882-1998.pdf

> Do you still want this #if'd?

My opinion is totally unimportant and irrelevant for getting this merged, but it seems like either `#if` or raise the required C++ version.

Reply:
Experimentally confirmed that `noexcept` requires c++11.  

Try this with various `-std` args:

```
#if __has_feature(cxx_noexcept)
#error yes noexcept
#else
#error no noexcept
#endif
```

The current version of the PR doesn't compile for me:

```
libtool: compile:  g++ -DPACKAGE_NAME=\"marisa\" -DPACKAGE_TARNAME=\"marisa\" -DPACKAGE_VERSION=\"0.2.6\" "-DPACKAGE_STRING=\"marisa 0.2.6\"" -DPACKAGE_BUGREPORT=\"susumu.yata@gmail.com\" -DPACKAGE_URL=\"\" -DPACKAGE=\"marisa\" -DVERSION=\"0.2.6\" -DHAVE_STDIO_H=1 -DHAVE_STDLIB_H=1 -DHAVE_STRING_H=1 -DHAVE_INTTYPES_H=1 -DHAVE_STDINT_H=1 -DHAVE_STRINGS_H=1 -DHAVE_SYS_STAT_H=1 -DHAVE_SYS_TYPES_H=1 -DHAVE_UNISTD_H=1 -DSTDC_HEADERS=1 -DHAVE_DLFCN_H=1 -DLT_OBJDIR=\".libs/\" -I. -march=native -Wall -Weffc++ -Wextra -Wconversion -I../../../../include -I../../../../lib -g -O2 -DMARISA_USE_POPCNT -mpopcnt -DMARISA_USE_SSE4 -msse4 -MT mapper.lo -MD -MP -MF .deps/mapper.Tpo -c mapper.cc  -fno-common -DPIC -o .libs/mapper.o
...
../../../../include/marisa/exception.h:22:11: error: exception specification of overriding function is more lax than base version
  virtual ~Exception() noexcept;
          ^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/exception:101:13: note: overridden virtual function is here
    virtual ~exception() _NOEXCEPT;
            ^
In file included from mapper.cc:13:
In file included from ../../../../lib/marisa/grimoire/io/mapper.h:6:
In file included from ../../../../include/marisa/base.h:191:
../../../../include/marisa/exception.h:22:23: error: expected ';' at end of declaration list
  virtual ~Exception() noexcept;
```

```
% g++ --version
Configured with: --prefix=/Applications/Xcode.app/Contents/Developer/usr --with-gxx-include-dir=/Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk/usr/include/c++/4.2.1
Apple clang version 12.0.0 (clang-1200.0.32.29)
Target: x86_64-apple-darwin19.6.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
```